compiler/executable/function.rs

```diff
@@ -75,3 +86,97 @@ fn compile_return_operation(
         }
     }
 }
+
+pub fn determine_tabling_requirements(
+    functions: &HashMap<FunctionID, &AnnotatedFunction>,
+) -> HashMap<FunctionID, bool> {
+    let mut cycle_detection: HashMap<FunctionID, TablingRequirement> =
+        functions.keys().map(|function_id| (function_id.clone(), TablingRequirement::Unexplored)).collect();
+    for (function_id, _) in functions {
+        if cycle_detection[function_id] == TablingRequirement::Unexplored {
+            determine_tabling_requirements_impl(functions, &mut cycle_detection, function_id)
+        }
+    }
+    cycle_detection
+        .into_iter()
+        .map(|(function_id, tabling_requirement)| match tabling_requirement {
+            TablingRequirement::KnownTabled => (function_id, true),
+            TablingRequirement::KnownUntabled => (function_id, false),
+            TablingRequirement::Unexplored | TablingRequirement::UnknownInStack => unreachable!(),
```

Comment:
Let's put descriptive messages into all `unreachable`s we have to share to context why it is so. Otherwise, it will be really tough to understand what went wrong when anybody else steps on it in 1 year.